source/renderer/app/components/widgets/forms/PinCode.js

```diff
@@ -31,42 +38,198 @@ export default class PinCode extends Component<Props> {
+            const inputFieldRef = this.inputsRef[focusKey];
+            if (inputFieldRef && inputFieldRef.inputElement) {
+              inputFieldRef.focus();
+              if (inputFieldRef.props.value) {
+                inputFieldRef.inputElement.current.selectionStart = 1;
+                inputFieldRef.inputElement.current.selectionEnd = 1;
+              }
+              this.setState({ focusKeyChanged: false });
+            }
+          }, 50);
+        } else {
+          newValue[key] = inputNewValue;
+          this.setState({ isBackSpace: false, focusKeyChanged: false });
+        }
+      }
       if (onChange) {
-        onChange(newValue);
+        onChange(newValue, inputNewValue, key);
       }
       this.focusKey = key;
-      this.add = inputValue !== null && inputValue !== undefined;
+      this.add =
+        inputValue !== null && inputValue !== undefined && !isNaN(inputValue);
     }
   };
 
   componentDidUpdate() {
-    const { value, length } = this.props;
+    const { value, length, name, selectedPinField } = this.props;
+    const { isBackSpace, focusKeyChanged } = this.state;
     const key = value.join('').length;
-    if (key > 0 && key < length) {
-      const inputFocusKey = this.add ? this.focusKey + 1 : this.focusKey - 1;
+    const inputValue = value[key - 1];
+    this.add = this.fromBackspace
+      ? false
+      : inputValue !== null && inputValue !== undefined && !isNaN(inputValue);
+    const emptyFieldIndex = value.findIndex((item) => item === '');
+    this.focusKey =
+      emptyFieldIndex !== undefined &&
+      emptyFieldIndex !== null &&
+      emptyFieldIndex > -1
+        ? emptyFieldIndex
+        : this.focusKey;
+    const focusKey = parseInt(this.focusKey, 10);
+    if (
+      name === selectedPinField &&
+      ((key > 0 && key < length) || emptyFieldIndex > -1 || focusKeyChanged)
+    ) {
+      let inputFocusKey = 0;
+      if (emptyFieldIndex > -1 && !this.fromBackspace) {
+        inputFocusKey = emptyFieldIndex;
+      } else {
+        inputFocusKey = this.add ? focusKey + 1 : focusKey - 1;
+      }
+      const inputElementRef = this.inputsRef[inputFocusKey];
       if (
         Object.prototype.hasOwnProperty.call(this.inputsRef, inputFocusKey) &&
-        this.inputsRef[inputFocusKey]
-      )
-        this.inputsRef[inputFocusKey].focus();
+        inputElementRef
+      ) {
+        this.fromBackspace = false;
+        if (
+          isBackSpace &&
+          inputElementRef.inputElement &&
+          emptyFieldIndex !== inputFocusKey
+        ) {
+          inputElementRef.focus();
+          if (inputElementRef.props.value) {
+            inputElementRef.inputElement.current.selectionStart = 1;
+            inputElementRef.inputElement.current.selectionEnd = 1;
+          }
+        } else if (!isBackSpace) {
+          inputElementRef.focus();
+        }
+      }
     }
   }
 
+  onKeyDown = (evt: SyntheticKeyboardEvent<EventTarget>, inputKey: number) => {
```

Comment:
Please add comments to the various edge cases that are handled here … nobody will know it better than you now and your/our future self will thank you for detailed comments about how/why certain checks are necessary

Reply:
@DominikGuzei Added comments with more descriptions on top of edge cases where input values where either added/deleted or focus was changed from field to field.